javagen/src/main/java/com/azure/autorest/template/util/ModelTemplateHeaderHelper.java

```diff
@@ -0,0 +1,257 @@
+        classBlock.publicConstructor(String.format("%s(HttpHeaders rawHeaders)", model.getName()), constructor -> {
+            // HeaderCollections need special handling as they may have multiple values that need to be retrieved from
+            // the raw headers.
+            List<ClientModelProperty> collectionProperties = new ArrayList<>();
+            for (ClientModelProperty property : model.getProperties()) {
+                if (CoreUtils.isNullOrEmpty(property.getHeaderCollectionPrefix())) {
+                    generateHeaderDeserializationFunction(property, constructor);
+                } else {
+                    collectionProperties.add(property);
+                }
+            }
+
+            if (!CoreUtils.isNullOrEmpty(collectionProperties)) {
+                // Bundle all collection properties into one iteration over the HttpHeaders.
+                generateHeaderCollectionDeserialization(collectionProperties, constructor);
+            }
+        });
+    }
+
+    /**
+     * Adds {@code private static final} {@link HttpHeaderName} constants representing the headers that are used by the
+     * {@link ClientModel}.
+     *
+     * @param classBlock The class block for the model.
+     * @param model The model itself.
+     */
+    private static void addHttpHeaderNameConstants(JavaClass classBlock, ClientModel model) {
+        for (ClientModelProperty property : model.getProperties()) {
+            if (!CoreUtils.isNullOrEmpty(property.getHeaderCollectionPrefix())) {
+                // Header collections aren't able to use HttpHeaderName.
+                continue;
+            }
+
+            if (HEADER_TO_KNOWN_HTTPHEADERNAME.containsKey(property.getSerializedName().toLowerCase(Locale.ROOT))) {
+                // Header is a well-known HttpHeaderName, don't need to create a private constant.
+                continue;
+            }
+
+            String headerName = property.getSerializedName();
+            String constantName = CodeNamer.getEnumMemberName(headerName);
+            classBlock.variable(
+                String.format("HttpHeaderName %s = HttpHeaderName.fromString(\"%s\")", constantName, headerName),
+                JavaVisibility.Private, JavaModifier.Static, JavaModifier.Final);
+        }
+    }
+
+    private static void generateHeaderDeserializationFunction(ClientModelProperty property, JavaBlock javaBlock) {
+        IType wireType = property.getWireType();
+        boolean needsNullGuarding = wireType != ClassType.String &&
+            (wireType instanceof ArrayType || wireType instanceof ClassType
+                || wireType instanceof EnumType || wireType instanceof GenericType);
+
+        // No matter the wire type the rawHeaders will need to be accessed.
+        String httpHeaderName = HEADER_TO_KNOWN_HTTPHEADERNAME.containsKey(property.getSerializedName().toLowerCase(Locale.ROOT))
+            ? "HttpHeaderName." + CodeNamer.getEnumMemberName(property.getSerializedName())
```

Comment:
Could there possible that the name of the property and the constant var name in `HttpHeaderName` is not connected via this `CodeNamer.getEnumMemberName` function?

Since you are doing reflection and created that `HEADER_TO_KNOWN_HTTPHEADERNAME` from `HttpHeaderName`, would `HEADER_TO_KNOWN_HTTPHEADERNAME.get(...)` gets the constant name in `HttpHeaderName`?